### ISSUE TITLE
`add_multiple_results.py` fails

### ISSUE THREAD

**Issue by bilby-bot** [bilby-dev/bilby#268]
In GitLab by @git.ligo:matthew-pitkin on Oct 9, 2018, 17:05

Using 47d29e5a9700713744349297d48c41437e53f60 installed under Python 3.5 I find that the `add_multiple_results.py` example fails. It seems that the `__add__` method of the `Results()` class is failing as it can't find a `samples` attribute in the class. Digging a bit deeper, somewhere between samples attribute being set [here](https://git.ligo.org/Monash/bilby/blob/master/bilby/core/sampler/emcee.py#L110) and the samplers being added the `samples` attribute is being deleted by [`samples_to_posterior()`](https://git.ligo.org/Monash/bilby/blob/master/bilby/core/result.py#L541).

Shoud the `__add__` method really be more like:

```python
def __add__(self, other):
    matches = ['sampler', 'search_parameter_keys']
    for match in matches:
        # The 1 and 0 here ensure that if either doesn't have a match for
        # some reason, a error will be thrown.
        if getattr(other, match, 1) != getattr(self, match, 0):
            raise ValueError(
                "Unable to add results generated with different {}".format(match))

    if not hasattr(self, 'samples') and not hasattr(other, 'samples'):
        if not hasattr(self, 'posterior') and not hasattr(other, 'posterior'):
            raise ValueError("Unable to add results")
        else:
            self.posterior = pd.concat([self.posterior, other.posterior])
    else:
        self.samples = np.concatenate([self.samples, other.samples])

    return self

```

**Comment by bilby-bot** [comment]
In GitLab by @git.ligo:matthew-pitkin on Oct 10, 2018, 09:55

In addition to fixing this, I think the `__add__` method should also be able to check whether nested sampling was used, and if so it should combine samples and recompute the posterior samples as in, e.g., [`draw_posterior_many()`](https://git.ligo.org/lscsoft/lalsuite/blob/master/lalinference/python/lalinference/nest2pos.py#L61) in LALSuite (and [cpnest](https://github.com/johnveitch/cpnest/blob/master/cpnest/nest2pos.py#L73)). cc @git.ligo:john-veitch. This should probably be a separate issues/MR though!

**Comment by bilby-bot** [comment]
In GitLab by @git.ligo:gregory.ashton on Oct 16, 2018, 09:34

@matthew\-pitkin good spot. This is an old example/code that hasn't been maintained. Indeed, I've actually proposed removing the `__add__` method in !241 . My suggestion is we actually remove it and add it back in in the manner you have suggested here.

**Comment by bilby-bot** [comment]
In GitLab by @git.ligo:matthew-pitkin on Oct 22, 2018, 10:53

@git.ligo:gregory.ashton cool, if I get the time after !241 is merged I might take a look into re-adding the `__add__` method as suggested, but I'd be happy for someone else to get there first. Once !241 is merged I might create a new issue about that specifically.

**Comment by bilby-bot** [comment]
In GitLab by @git.ligo:gregory.ashton on Jan 8, 2019, 00:09

@matthew\-pitkin , I'm going to close this now as we no longer have an `__add__`, feel free to open a new issue to add them back in properly :)

**Comment by bilby-bot** [comment]
In GitLab by @git.ligo:gregory.ashton on Jan 8, 2019, 00:09

closed

**Comment by bilby-bot** [comment]
In GitLab by @git.ligo:michael.williams on Oct 3, 2024, 17:56

unassigned @git.ligo:moritz.huebner
